Remmeauth/remme-core
# Atomic swap

Atomic swaps inplementation for REMChain <=> OTHER_CHAIN swaps:

- _swap_init
- _swap_approve
- _swap_expire
- _swap_set_lock
- _swap_close

IN THE PROCESS CHANGES MADE TO:
- added Settings Handler 
- refactored hashlib to a single point of reference location
- package naming for convenient tp lookup
- work with proper address formation
- refactoring special addresses: ZERO_ADDRESS, GENESIS_ADDRESS
- unified testing suit and handlers actions for token_transfers
- open api and payload for each method
- minor bugs

UNITTEST ISSUE:
https://remmeio.atlassian.net/wiki/spaces/WikiREMME/pages/81952778/TEMPORARY+NOTES+FOR+ATOMIC+SWAPS

CURRENT PROGRESS OF TEST COVERAGE:
- Successful init ETH => REMchain (DONE) 
- Successful init REMchain => ETH 
- Successful close (DONE) 
- Fail to close 
- Fail to expire 
- Fail to set lock
- Fail to approve 

TO BE ADDED:
- input validation

REST API UI:
http://localhost:8080/api/v1/ui/

TO GET THE LATEST TIMESTAMP FOR "created_at" FIELD
https://www.unixtimestamp.com/

MANUAL TESTING OF THE REST API
1. ETH => REMchain (Bob Actions)
```
1. /init/
{
  "amount": 10000,
  "created_at": 1526293647,
  "email_address_encrypted_optional_alice": "",
  "is_initiator": false,
  "receiver_address": "2265da5c1a648002310b7360b38d82e0b6357e42d49fb1d6b8b33b2fad7a659f2d4e44",
  "secret_lock_optional_bob": "8bc9d977e4d5ef40f1a61fb60b687eb1924cd90ed4061f31923f36a057f54ce5",
  "sender_address_non_local": "",
  "swap_id": "033102e41346242476b15a3a7966eb5249271025fc7fb0b37ed3fdb4bcce482c"
}

2. /expire/
Expected behaviour: asked to wait 48 hours
{
  "swap_id": "033102e41346242476b15a3a7966eb5249271025fc7fb0b37ed3fdb4bcce482c"
}
3. /close/
{
  "secret_key": "747f9238d844e6ae6239df18716bb3dc486e6a9dee7ed6630f8aa6f05946440b",
  "swap_id": "033102e41346242476b15a3a7966eb5249271025fc7fb0b37ed3fdb4bcce482c"
}
```
2. REMchain => ETH (Alice Actions)
```
1. /init/
{
  "amount": 10000,
  "created_at": 1526391085,
  "email_address_encrypted_optional_alice": "",
  "is_initiator": false,
  "receiver_address": "2265da5c1a648002310b7360b38d82e0b6357e42d49fb1d6b8b33b2fad7a659f2d4e44",
  "sender_address_non_local": "",
  "swap_id": "033102e41346242476b15a3a7966eb5249271025fc7fb0b37ed3fdb4bcce482c"
}

2. /expire/
Expected behaviour: wait for 24 hours
{
  "swap_id": "033102e41346242476b15a3a7966eb5249271025fc7fb0b37ed3fdb4bcce482c"
}

3. /set-secret-lock/
{
  "secret_lock": "8bc9d977e4d5ef40f1a61fb60b687eb1924cd90ed4061f31923f36a057f54ce5",
  "swap_id": "033102e41346242476b15a3a7966eb5249271025fc7fb0b37ed3fdb4bcce482c"
}

4. /approve/
{
  "swap_id": "033102e41346242476b15a3a7966eb5249271025fc7fb0b37ed3fdb4bcce482c"
}

5. /close/
{
  "secret_key": "747f9238d844e6ae6239df18716bb3dc486e6a9dee7ed6630f8aa6f05946440b",
  "swap_id": "033102e41346242476b15a3a7966eb5249271025fc7fb0b37ed3fdb4bcce482c"
}
```

{
  "secret_key": "747f9238d844e6ae6239df18716bb3dc486e6a9dee7ed6630f8aa6f05946440b",
  "swap_id": "033102e41346242476b15a3a7966eb5249271025fc7fb0b37ed3fdb4bcce482c"
}
